test/test_helper/common.bash

```diff
@@ -210,30 +210,52 @@ function wait_for_empty_mail_queue_in_container() {
 # Common defaults appropriate for most tests, override vars in each test when necessary.
 # TODO: Check how many tests need write access. Consider using `docker create` + `docker cp` for easier cleanup.
 function init_with_defaults() {
+  export TEST_NAME TEST_TMP_CONFIG
+
   # Ignore absolute dir path and file extension, only extract filename:
-  export TEST_NAME
   TEST_NAME="$(basename "${BATS_TEST_FILENAME}" '.bats')"
+  TEST_TMP_CONFIG="$(duplicate_config_for_container . "${TEST_NAME}")"
+
+  export TEST_FILES_CONTAINER_PATH='/tmp/docker-mailserver-test'
+  export TEST_FILES_VOLUME="${PWD}/test/test-files:${TEST_FILES_CONTAINER_PATH}:ro"
 
-  export PRIVATE_CONFIG
-  PRIVATE_CONFIG="$(duplicate_config_for_container . "${TEST_NAME}")"
-  export TEST_FILES_VOLUME="${PWD}/test/test-files:/tmp/docker-mailserver-test:ro"
-  export TEST_CONFIG_VOLUME="${PRIVATE_CONFIG}:/tmp/docker-mailserver:ro"
+  # The config volume cannot be read-only as some data needs to be written at container startup
+  # - two sed failures (unknown lines)
+  # - dovecot-quotas.cf (setup-stack.sh:_setup_dovecot_quotas)
+  # - postfix-aliases.cf (setup-stack.sh:_setup_postfix_aliases)
+  export TEST_CONFIG_VOLUME="${TEST_TMP_CONFIG}:/tmp/docker-mailserver"
 
   export TEST_FQDN='mail.my-domain.com'
+  export TEST_DOCKER_ARGS
+
+  export TEST_CA_CERT="${TEST_FILES_CONTAINER_PATH}/ssl/example.test/with_ca/ecdsa/ca-cert.ecdsa.pem"
 }
 
-# Common docker run command that should satisfy most tests which only modify ENV.
+# Using `create` and `start` instead of only `run` allows to modify
+# the container prior to starting it. Otherwise use this combined method.
 function common_container_setup() {
-  local TEST_ENV_FILE=$1
+  common_container_create "$@"
+  common_container_start
+}
 
-  run docker run -d --name "${TEST_NAME}" \
-    --volume "${TEST_FILES_VOLUME}" \
-    --volume "${TEST_CONFIG_VOLUME}" \
+# Common docker setup is centralized here,
+# can pass an array as a reference for adding extra config
+function common_container_create() {
+  local -n X_EXTRA_ARGS=${1}
```

Comment:
What does `local -n` do exactly? Never seen this before. Does ot create an array here (I assume because the variable is used as such later)?

Reply:
`help local` gives this:

`      -n        make NAME a reference to the variable named by its value`

Reply:
Yeah, I needed a way to pass the array over to the method. I see that I also export `TEST_DOCKER_ARGS` in a related `common.bash` function while doing a `local` declaration in the test.

There was probably a point to that, maybe with the updated `letsencrypt` test which required the updates made in this PR. :thinking:

Reply:
But it's good we're talking about it; as we can see, there's a bit of confusion about this:) There should be a short comment explaining what's going on :D

Reply:
I don't understand what exactly `local -n X_EXTRA_ARGS=${1}` try to solve her. Why is a named reference needed?

Reply:
> Why is a named reference needed?

`test => ( (method1 => method2) ||  method2 ) => X_EXTRA_ARGS[@]`

It just allows a single arg/var to expand where desired while allowing adding other args and not having to deal with that mess if just passing `$@` and changing the function to work with multiple inputs later?

I'd like to convert all the tests eventually to share the common config, but some like the other SSL tests have some more involved setups, and I'd like some flexibility in configuring a test matrix run (eg run the test suite with a different set of defaults like `example.com` vs `mail.example.com` without having to manually change / alter all tests individually.

So it's mostly for convenience while iterating on it.

Reply:
I am still not getting it 100%. If you like, you can give a generic code example, maybe that enlightens me 😆 All what you wrote seem to me perfectly doable without named references. That's what confusing me right know. Also I don't understand how this actually works.

Here, an array with two elements is defined:

https://github.com/polarathene/docker-mailserver/blob/d8644fbb09bfaaa3dec03fc611646adcbd67e140/test/mail_spam_bounced.bats#L53


Here, function common_container_setup is called with a string argument 'TEST_DOCKER_ARGS'
https://github.com/polarathene/docker-mailserver/blob/d8644fbb09bfaaa3dec03fc611646adcbd67e140/test/mail_spam_bounced.bats#L57

common_container_setup passes the string 'TEST_DOCKER_ARGS' to the function common_container_create.
https://github.com/polarathene/docker-mailserver/blob/d8644fbb09bfaaa3dec03fc611646adcbd67e140/test/test_helper/common.bash#L248


Assumption: Here some magic happens and X_EXTRA_ARGS now contains the two elements from defined array defined at the beginning.
https://github.com/polarathene/docker-mailserver/blob/d8644fbb09bfaaa3dec03fc611646adcbd67e140/test/test_helper/common.bash#L260

I am wondering, how a function can access other local variables.

Still not understanding, why in that case `local -n X_EXTRA_ARGS=${1}` is needed and the docker run command is not just using `"$@" \`, instead of `"${X_EXTRA_ARGS[@]}" \`

Thank you for your patience 👍

Reply:
> Assumption: Here some magic happens and X_EXTRA_ARGS now contains the two elements from defined array defined at the beginning.

```bash
local -n X_EXTRA_ARGS=${1}
```

- Take the 1st parameter received (a string `TEST_DOCKER_ARGS`) and via `-n` reference the variable from the current scope.
- `local` scopes it the declaration/value within the bounds of the function with the `local` keyword, it's value is visible to sub-function calls until the function with the `local` scope has finished and the value is lost or changed back to what it was previously defined as.
- When the BATS tests run, IIRC, `setup_file()` is run once, and can use `export` to make variables available to the individual test cases. `local` will not pass over to these, I'm not sure what's going on under the hood, BATS perhaps is calling them as subprocesses/subshells or something where `export` works? While the test itself runs (which is scoped within the `@test "description" { ... }`, it can define `local` and regular bash functions will be called afterwards so scoping works fine without using `export`.
- Any functions called that define their own `local` keyworded vars will also have those limited scope to those functions bounds, thus not be available to anything further up the call chain.
- I chose the name `X_EXTRA_ARGS` here instead to try emphasize it was temporary, which although the `local` probably does that, this var is working a bit differently as a reference with `-n` so I chose that `X_` prefix convention :sweat_smile: 
- I don't think I could provide the same `TEST_DOCKER_ARGS` var name as it might not work well with the reference feature (I haven't confirmed that), and I'm not requiring that the test use the array var name `TEST_DOCKER_ARGS` (_otherwise we could probably avoid parsing in a parameter at all and hard-code the string I guess?_), the `X_` prefix also sort of safeguards against someone using an variable `EXTRA_ARGS` which might not be passed in as a string but if it would conflict, this was just an easy way to not worry about it :sweat_smile:
- Anything prefixed with `TEST_` here is usually intended for these `common_container` methods and can be overrided at any stage of the test where needed (in the letsencrypt test PR open atm, I do that with `TEST_CA` for one test case.

> Still not understanding, why in that case `local -n X_EXTRA_ARGS=${1}` is needed and the docker run command is not just using `"$@" \`, instead of `"${X_EXTRA_ARGS[@]}" \`

Presently that'd require each test case calling `common_container_setup "${TEST_DOCKER_ARGS[@]}"` as well. Then passing that through the method to `common_container_create` (unless calling this directly instead), and then expanding with just `"$@" as you mention.

The current approach is more terse in that sense, and I kind of prefer seeing the descriptive `X_EXTRA_ARGS` as an arg where it's being used, even if that might still be obvious what it's doing with just `"$@" \` :man_shrugging: 

More importantly though, what happens if you want to have additional parameters for `common_container_setup` or `common_container_create` that aren't these array of values to expand in the `docker create` command? Then it seems to get more messy to handle, vs defining a new local var to catch `${2}` or whatever and using that where needed.

There is a lot of tests to adapt to these helper methods, I'm not sure yet if just an array of values for `docker create` will be sufficient. Hopefully that clarifies things better :)

Reply:
> Hopefully that clarifies things better :)

Absolutely, thank you! Especially `local TEST_DOCKER_ARGS` confused me, cause of the `local`.

> it's value is visible to sub-function calls

Didn't know that. I thought, this was only for a function and not sub-function calls. But it makes sense.

Reply:
I removed the `export` in `common.bash:init_with_defaults()` for `TEST_DOCKER_ARGS` with the previous commit since I don't think I have a need for it atm, it was probably for declaring common extra args for an entire test file in `setup_file()` but as it's not referenced anywhere globally and just passed as a reference var to `common_container_setup()`, that could just be exported in `setup_file()` when needed :sweat_smile: 

If you think I could document anything better let me know :+1: